Add the ability to destroy hosts from the front-end

jonc125 opened the pull request:
This will fix #30 when complete.

Some issues outstanding:
- [x] What happens if we destroy a host that's being deployed?
- [ ] Do we need explicit different buttons for 'delete' and 'destroy' (and better names for them)?
- [ ] Should we do #29 as part of this PR too?
- [x] Will probably need better error handling, c.f. #15, but maybe as a subsequent PR?
- [ ] It's irritating that we have to render the host config and initialise Terraform before we can destroy a host! Might be nice to figure out how to do a global `terraform init` rather than on every operation.

jonc125 committed:
Make self.tfstate_path actually be the state file path

jonc125 committed:
Destroy hosts from the web front-end

This adds a new `destroy_host` method to the Destroyer, which uses
the `run_cmd` approach to capture the destroy call output. It also
adds an extra host status for use while the destroy is happening.

There's a little bit more error handling for the Celery tasks, so
exceptions get reported in the deploy log, but this will need to be
more robust.

The 'delete' button now does a destroy, but leaves the host in
'defining' mode. Clicking delete again will remove it from the DB.

jonc125 commented:
This PR is definitely still work-in-progress, but might be nice to show in the demo @raquel-ucl ? If we merge the other ones, I can retarget this and just do the demo off this branch perhaps?

jonc125 committed:
Fix #39, removing use of python-terraform

Much of the `destroy` and `refresh` methods was not actually used,
so I've just removed it, and renamed `destroy_host` to `destroy`
for simplicity.

(Not tested locally - will check what Travis thinks!)

ageorgou commented:
I think it makes sense to continue this as part of implementing additional operations (see also #9).
@jonc, was there anything problematic with this, off the top of your head?

jonc125 commented:
Not beyond what I noted in the PR description above, IIRC.

Commit:
Merge pull request #45 from UCL-CloudLabs/remove-python-terraform

Remove use of python-terraform

ageorgou committed:
Add placeholder for getting status

ageorgou committed:
Add initial plan for getting status

This is quite hacky and I don't like it, but it should work. Getting
the status on Azure is not straightforward...

ageorgou committed:
Finish a rough version of getting status

ageorgou committed:
Reorder migrations to not confuse alembic

The DB migrations had branched between the naming changes and the
ones from this series of changes. Let's just place the replacement
of the enum after the naming changes.

ageorgou committed:
Display 'live' status instead of checking the DB

ageorgou committed:
Update requirements with Azure SDK

ageorgou committed:
Improve PEP8 compliance

ageorgou committed:
Remove debug prints, oops

ageorgou committed:
Move Azure-related functionality to its own module

ageorgou committed:
Add celery task for stopping a VM

ageorgou committed:
Clarify that Azure 'stopped' status is not yet off

A stopped (but not deallocated) VM still incurs charges, whereas a
dellocated one does not. It would therefore be misleading to show
a stopped machine as actually stopped, especially since the status
could then move back to 'stopping' (while deallocating). I am not
sure whether deallocating happens as a second step or is a wholly
separate operation, but in either case I think it should be fine
to treat Azure 'stopped' as a temporary status, i.e. map it to our
'stopping' Host status.

ageorgou committed:
Launch stop task from dashboard

ageorgou committed:
Add task for starting a stopped VM

ageorgou committed:
Merge branch 'development' into destroy

ageorgou committed:
Add status for DB entries not found in the cloud

In case e.g. a host has been destroyed, but hasn't been deleted
from the database yet. Currently only used when displaying the
status in the dashboard. Fixes #57.

ageorgou committed:
Update the DB with each VM's live status

Added a periodic task that queries Azure and updates the existing
hosts' statuses in the DB. We will need to run `celery beat` for
this, in addition to starting the Celery workers. At the moment,
the dashboard displays the live status but the buttons depend on
the DB status (which will be updated eventually, anyway). Ideally,
we want some logic to combine the live status with info from the
application, particularly to identify hosts being deployed (which
we cannot do with information from Azure only).

ageorgou committed:
Ignore live status for hosts still in deployment

ageorgou committed:
Show database status on dashboard

Now that the databse is updated in the background, we can show that
status on the dashboard instead of the live one.

ageorgou committed:
Store Terraform template in database

ageorgou committed:
Create list of current deployments

ageorgou committed:
Delete deploying hosts through Azure SDK

ageorgou committed:
Rerender template if not found in DB when deleting

raquelalegre committed:
Update to latest version of Azure's TF plugin.

raquelalegre committed:
Azure changed its way to define a vnet and it is now mandatory to include a reference from the vnet definition to the subnet.

ageorgou committed:
Remove empty line

ageorgou committed:
Merge branch 'bugfix/terraform_vnet' into destroy-manage

ageorgou committed:
Store task ID of new deployments in DB

ageorgou committed:
Add the stopping host status (somehow missed!)

ageorgou committed:
Comment on new celery details

ageorgou committed:
Clean up redundant code

As per PR comments. I've left in the debug messages as comments for
when we implement logging. Also removing another unused import.

ageorgou committed:
Capture Azure functionality in a class

ageorgou committed:
Use new Azure tools class

This makes some of the deployer methods (except for deploy and
destroy) trivial. They could probably be removed and we could
then just have the start, stop etc methods called form the celery
tasks. I am leaving them in for now because we may still want to
do something for which we need the deployer (eg update the DB).
This commit also makes the start/restart methods more consistent
(restart is only for hosts that are already running, start is for
either stopped or running).

ageorgou committed:
Move status retrieval to new tools class

Plus small cleanup and fixes.

ageorgou committed:
Move "deployment unsetting" to right place!

Not sure how that was missed -- we are meant to unset the task
field after a (successful or failed) deployment attempt, not after
a reset!

ageorgou committed:
Enable starting a stopped machine functionality

Previously only restarting and stopping were available.

ageorgou committed:
Merge pull request #58 from UCL-CloudLabs/destroy-manage

Add other actions on VMs from the dashboard
